### PR TITLE
Dev: refine non-root sudoer support for crmsh.parallax (bsc#1210709)

### PR DESCRIPTION
**Pull request by nicholasyang2022**
# Problem

The current implementation of non-root sudoer support of crmsh.parallax is hardcoded `~<sudoer>/.ssh/id_rsa` as the identity file for creating ssh sessions. When the sudoer uses a non-RSA key, or specifies an alternative path in configurations, it will fails to create an ssh session.

Moreover, some direct calls to `parallax` module, without going through `crmsh.parallax`, still exist in the code. They do not work in non-root scenarios.

# Solution

1. Create a new concurrent ssh util that supports non-root sudoer natively at `crmsh.prun`. It su to the sudoer before creating ssh connections and sudo to root on the remote host once the connections are established and authenticated. Thanks to the new `asyncio` module in python 3, this new util is much more simpler than parallax.
2. Rewrite `crmsh.parallax` with `crmsh.prun`.
3. Replace those direct calls to `parallax` module with `crmsh.parallax` or `crmsh.prun`.
4. Remove dependency `parallax` from `setup.py`.

### CODE REVIEW

**Comment by liangxin1300** [comment]
Issue with -x option: still require password, even already configured password-less under user `xin`
```
xin@15sp4-1:~> sudo crm cluster init -y -x
INFO: Loading "default" profile from /etc/crm/profiles.yml
WARNING: chronyd.service is not configured to start at system boot.
INFO: SSH key for hacluster does not exist, hence generate it now
INFO: Configuring corosync (unicast)
WARNING: Not configuring SBD - STONITH will be disabled.
INFO: Hawk cluster interface is now running. To see cluster status, open:
INFO:   https://192.168.122.192:7630/
INFO: Log in with username 'hacluster', password 'linux'
WARNING: You should change the hacluster password to something more secure!
INFO: BEGIN Waiting for cluster
...........                                                                                                                   INFO: END Waiting for cluster
INFO: Loading initial cluster configuration
INFO: Done (log saved to /var/log/crmsh/crmsh.log)
```

```
xin@15sp4-2:~> sudo crm cluster join -c 15sp4-1 -y
WARNING: chronyd.service is not configured to start at system boot.
INFO: SSH key for hacluster does not exist, hence generate it now
WARNING: Might have unfinished process on other nodes, wait 120s...
INFO: BEGIN Retrieve all config files
Password: 
Password: 
Password:
```

File: crmsh/prun/prun.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+def _is_local_host(host):
+    """
+    Check if the host is local
+    """
+    try:
+        socket.inet_aton(host)
+        hostname = socket.gethostbyaddr(host)[0]
+    except OSError:
+        hostname = host
+    return hostname == socket.gethostname()
```
Comment (liangxin1300) [inline crmsh/prun/prun.py:250]: Please add more descriptions to introduce this .py file

**Comment by liangxin1300** [comment]
Python3 version in 15SP5 is 3.6.15, please double check that we don't use the interfaces of asyncio that belong to newer python version

**Comment by nicholasyang2022** [comment]
> Issue with -x option: still require password, even already configured password-less under user xin

The `scp` in `retrieve_all_config_files` has been replaced with with `ssh cpio` and it works with non-root sudoer now.

And more comments have been added to `runner.py` and `prun.py`.

> Python3 version in 15SP5 is 3.6.15, please double check that we don't use the interfaces of asyncio that belong to newer python version

I used the API docs of python 3.6.12 as reference when I wrote these code. It is unlikely to use a API not available in 3.6.

**Comment by liangxin1300** [comment]
A regression:
```
15sp4-1:~ # crm configure primitive d Dummy
15sp4-1:~ # crm resource stop d
15sp4-1:~ # crm configure rsctest d
Probing resources .
testing on 15sp4-1: d
testing on 15sp4-2: d
```
Then applied this PR:
```
15sp4-1:~ # crm configure rsctest d
Probing resources .Traceback (most recent call last):
  File "/usr/sbin/crm", line 52, in <module>
    rc = main.run()
  File "/usr/lib/python3.6/site-packages/crmsh/main.py", line 369, in run
    return main_input_loop(context, user_args)
  File "/usr/lib/python3.6/site-packages/crmsh/main.py", line 247, in main_input_loop
    rc = handle_noninteractive_use(context, user_args)
  File "/usr/lib/python3.6/site-packages/crmsh/main.py", line 204, in handle_noninteractive_use
    if context.run(' '.join(l)):
  File "/usr/lib/python3.6/site-packages/crmsh/ui_context.py", line 86, in run
    rv = self.execute_command() is not False
  File "/usr/lib/python3.6/site-packages/crmsh/ui_context.py", line 271, in execute_command
    rv = self.command_info.function(*arglist)
  File "/usr/lib/python3.6/site-packages/crmsh/ui_configure.py", line 1228, in do_rsctest
    return rsctest.test_resources(rsc_l, node_l, all_nodes)
  File "/usr/lib/python3.6/site-packages/crmsh/rsctest.py", line 446, in test_resources
    if not are_all_stopped(resources, all_nodes):
  File "/usr/lib/python3.6/site-packages/crmsh/rsctest.py", line 409, in are_all_stopped
    drv.probe()
  File "/usr/lib/python3.6/site-packages/crmsh/rsctest.py", line 220, in probe
    self.runop("probe")
  File "/usr/lib/python3.6/site-packages/crmsh/rsctest.py", line 189, in runop
    self.ec_l[node] = results[node].returncode
TypeError: list indices must be integers or slices, not str
```

I will create a new PR to complete related behave test cases

**Comment by liangxin1300** [comment]
A regression for `crm script`:
```
# on a 2-node running cluster
# crm script run health
INFO: Verify health and configuration
INFO: Nodes: hanode1, hanode2
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/crmsh/scripts.py", line 2045, in run
    runner.prepare(has_remote_actions)
  File "/usr/lib/python3.6/site-packages/crmsh/scripts.py", line 1734, in prepare
    _create_remote_workdirs(self.printer, self.hosts, self.workdir, self.timeout_seconds)
  File "/usr/lib/python3.6/site-packages/crmsh/scripts.py", line 1599, in _create_remote_workdirs
    timeout_seconds).items():
  File "/usr/lib/python3.6/site-packages/crmsh/scripts.py", line 1078, in _parallax_call
    return crmsh.parallax.parallax_call(hosts, cmd, timeout_seconds=timeout_seconds)
  File "/usr/lib/python3.6/site-packages/crmsh/parallax.py", line 19, in parallax_call
    results = prun.prun({node: cmd for node in nodes}, timeout_seconds=timeout_seconds)
  File "/usr/lib/python3.6/site-packages/crmsh/prun/prun.py", line 69, in prun
    tasks = [_build_run_task(host, cmdline) for host, cmdline in host_cmdline.items()]
  File "/usr/lib/python3.6/site-packages/crmsh/prun/prun.py", line 69, in <listcomp>
    tasks = [_build_run_task(host, cmdline) for host, cmdline in host_cmdline.items()]
  File "/usr/lib/python3.6/site-packages/crmsh/prun/prun.py", line 99, in _build_run_task
    local_sudoer, remote_sudoer = crmsh.utils.user_pair_for_ssh(remote)
  File "/usr/lib/python3.6/site-packages/crmsh/utils.py", line 222, in user_pair_for_ssh
    return _user_of_host_instance.user_pair_for_ssh(host)
  File "/usr/lib/python3.6/site-packages/crmsh/utils.py", line 140, in user_pair_for_ssh
    remote_user = self.user_of(host)
  File "/usr/lib/python3.6/site-packages/crmsh/utils.py", line 125, in user_of
    ret = self._get_user_of_host_from_config(host)
  File "/usr/lib/python3.6/site-packages/crmsh/utils.py", line 163, in _get_user_of_host_from_config
    canonical, aliases, _ = socket.gethostbyaddr(host)
TypeError: gethostbyaddr() argument 1 must be str, bytes or bytearray, not tuple
```

**Comment by nicholasyang2022** [comment]
I have fix the regressions in

1. `cluster geo_join`
2. `configure rsctest`
3. `script`

**Comment by liangxin1300** [comment]
> I have fix the regressions in
> 
> 1. `cluster geo_join`
> 2. `configure rsctest`
> 3. `script`

Verified in #1179, thanks!
I will take a look at details recently

**Comment by liangxin1300** [comment]
Original:
```
# crm cluster run "sh /opt/test.sh"
INFO: [15sp4-1]
ERROR: [15sp4-3]: Exited with error code 1
ERROR: [15sp4-2]: Exited with error code 127, Error output: sh: /opt/test.sh: No such file or directory
```
This PR:
```
 # crm cluster run "sh /opt/test.sh"
INFO: [15sp4-1]
ERROR: [15sp4-2]: rc=127

sh: /opt/test.sh: No such file or directory

ERROR: [15sp4-3]: rc=1
```
I prefer the previous format.

**Comment by liangxin1300** [comment]
Although current code for this `checksum` was also broken, I got this error
```
 # crm -d cluster diff "/etc/corosync/corosync.conf" --checksum
DEBUG: 2023/05/08 22:26:15
DEBUG: crm(live/15sp4-1)# cluster diff /etc/corosync/corosync.conf --checksum
Host              SHA1 checksum of /etc/corosync/corosync.conf
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/crmsh/ui_context.py", line 86, in run
    rv = self.execute_command() is not False
  File "/usr/lib/python3.6/site-packages/crmsh/ui_context.py", line 271, in execute_command
    rv = self.command_info.function(*arglist)
  File "/usr/lib/python3.6/site-packages/crmsh/ui_cluster.py", line 750, in do_diff
    utils.remote_checksum(filename, nodes, this_node)
  File "/usr/lib/python3.6/site-packages/crmsh/utils.py", line 2189, in remote_checksum
    _, _, _, path = result
ValueError: too many values to unpack (expected 4)
ERROR: cluster.diff: too many values to unpack (expected 4)
```

**Comment by liangxin1300** [comment]
For `crm cluster copy`:

Original:
```
# crm cluster copy "/opt/txt.1"
ERROR: Failed to push /opt/txt.1 to 15sp4-3: Exited with error code 1, Error output: lost connection

INFO: 15sp4-2
```
This PR:
```
# crm cluster copy "/opt/txt.1"
ERROR: Failed to copy /opt/txt.1 to root@15sp4-3:Cannot create SSH connection to root@15sp4-3: ssh: connect to host 15sp4-3 port 22: Connection refused
Connection closed.  
Connection closed
```

**Comment by liangxin1300** [comment]
For `crm cluster copy`:

Original:
```
# crm cluster copy "/opt/txt.1"
ERROR: Failed to push /opt/txt.1 to 15sp4-3: Exited with error code 1, Error output: lost connection

INFO: 15sp4-2
```
This PR:
```
# crm cluster copy "/opt/txt.1"
ERROR: Failed to copy /opt/txt.1 to root@15sp4-3:Cannot create SSH connection to root@15sp4-3: ssh: connect to host 15sp4-3 port 22: Connection refused
Connection closed.  
Connection closed
```

**Comment by liangxin1300** [comment]
Remove parallax from spec file and dockerfile https://github.com/ClusterLabs/crmsh/pull/1179/commits/eef7c5bf7a7d179194ef0f98a8fe96d05c76a2e5

**Comment by liangxin1300** [comment]
CI docker image already updated, to void errors, please merge https://github.com/ClusterLabs/crmsh/pull/1180 first

File: crmsh/parallax.py [hunk 2]
```diff
@@ -1,110 +1,31 @@
...
+    for node, result in results.items():
+        if isinstance(result, prun.PRunError):
+            raise ValueError('Failed to run command {} on {}@{}: {}'.format(cmd, result.user, result.host, result))
+        elif result.returncode != 0:
+            raise ValueError("Failed on {}: {}".format(node, crmsh.utils.to_ascii(result.stderr)))
+    return [(node, (result.returncode, result.stdout, result.stderr)) for node, result in results.items()]
 
 
-def parallax_slurp(nodes, localdir, filename, askpass=False, ssh_options=None, strict=True):
+def parallax_slurp(nodes: typing.Sequence[str], localdir, filename, askpass=False, ssh_options=None, strict=True) -> typing.List[typing.Tuple[str, typing.Union[str, Error]]]:
```
Comment (liangxin1300) [inline crmsh/parallax.py:28]: Some parameters do not appear to be used and should be removed.

File: crmsh/parallax.py [hunk 3]
```diff
@@ -114,12 +35,14 @@ def parallax_slurp(nodes, localdir, filename, askpass=False, ssh_options=None, s
     ssh_options: Extra options to pass to SSH
     Returns [(host, (rc, stdout, stdin, localpath)), ...] or ValueError exception
     """
```
Comment (liangxin1300) [inline crmsh/parallax.py:37]: Then the docstring should be updated

File: crmsh/parallax.py [hunk 4]
```diff
@@ -1,110 +1,31 @@
...
-        cmd = f'sudo bash -c "{self.cmd}"' if sudoer else self.cmd
-        host_port_user = []
-        for host in self.nodes:
-            _, remote_user = crmsh.utils.user_pair_for_ssh(host)
-            host_port_user.append([host, None, remote_user])
-        return parallax.run(host_port_user, cmd, self.opts)
-
-
-def parallax_call(nodes, cmd, askpass=False, ssh_options=None, strict=True):
+def parallax_call(nodes, cmd, *, timeout_seconds: int = -1):
```
Comment (liangxin1300) [inline crmsh/parallax.py:12]: What does this `*` mean?

File: crmsh/prun/prun.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+        return task
+    def result(self, result: ProcessResult) -> ProcessResult:
+        return result
+    def exception(self, exc: PRunError) -> PRunError:
+        return exc
+
+
+def prun(
+        host_cmdline: typing.Mapping[str, str],
+        *,
```
Comment (liangxin1300) [inline crmsh/prun/prun.py:52]: Also here, add a meaningful name for this `*`, or remove it if no use

File: crmsh/prun/prun.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    for task in tasks:
+        task = interceptor.task(task)
+        runner.add_task(task)
+    runner.run(timeout_seconds)
+    return {task.context['host']: _handle_run_result(task, interceptor) for task in tasks}
+
+
+def prun_multimap(
+        host_cmdline: typing.Sequence[typing.Tuple[str, str]],
+        *,
```
Comment (liangxin1300) [inline crmsh/prun/prun.py:80]: also here

File: crmsh/prun/prun.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+        return interceptor.exception(SSHError(task.context['ssh_user'], task.context['host'], crmsh.utils.to_ascii(task.stderr)))
+    else:
+        return interceptor.result(ProcessResult(task.returncode, task.stdout, task.stderr))
+
+
+def pcopy_to_remote(
+        src: str,
+        hosts: typing.Sequence[str], dst: str,
+        recursive: bool = False,
+        *,
```
Comment (liangxin1300) [inline crmsh/prun/prun.py:137]: also here

File: crmsh/parallax.py [hunk 8]
```diff
@@ -1,110 +1,31 @@
...
-        sudoer = userdir.get_sudoer()
-        cmd = f'sudo bash -c "{self.cmd}"' if sudoer else self.cmd
-        host_port_user = []
-        for host in self.nodes:
-            _, remote_user = crmsh.utils.user_pair_for_ssh(host)
-            host_port_user.append([host, None, remote_user])
-        return parallax.run(host_port_user, cmd, self.opts)
-
-
-def parallax_call(nodes, cmd, askpass=False, ssh_options=None, strict=True):
```
Comment (liangxin1300) [inline crmsh/parallax.py:94]: What if caller what to add other ssh options not only the default `crmsh.constants.SSH_OPTION`?
I suggest leave a parallax for custom ssh_option

File: crmsh/parallax.py (hunk 4 above)
Comment (nicholasyang2022) [reply crmsh/parallax.py:12]: It is PEP-3102 *Keyword-only Arguments*.

File: crmsh/parallax.py (hunk 8 above)
Comment (nicholasyang2022) [reply crmsh/parallax.py:94]: This parameter is unused. I prefer to add the parameter when it is really needed.

**Review by liangxin1300 — APPROVED** [review]
Great work. Thanks!

**Pull request merged**